client/src/features/visualizer-vivagraph/hooks/useFeed.ts

```diff
@@ -232,12 +236,61 @@ export const useFeed = (network: string) => {
             const selectedColor = getBlockColorByState(themeMode, metadataUpdate.blockState);
 
             updateBlockIdToMetadata(metadataUpdate.blockId, { color: selectedColor });
-            updateBlockColor(metadataUpdate.blockId, selectedColor);
+
+            const blockMetadata = getBlockIdToMetadata(metadataUpdate.blockId);
+            if (blockMetadata) {
+                const previousBlockState = blockMetadata.state;
+                const wasConfirmedBeforeAccepted = previousBlockState === "accepted" && metadataUpdate.blockState === "confirmed";
+
+                if (!wasConfirmedBeforeAccepted) {
+                    updateBlockIdToMetadata(metadataUpdate.blockId, {
+                        state: metadataUpdate.blockState,
+                    });
+                }
+
+                updateBlockColor(metadataUpdate.blockId, selectedColor);
+                const acceptedStates: BlockState[] = ["confirmed", "accepted"];
+                if (acceptedStates.includes(metadataUpdate.blockState)) {
+                    const slot = Utils.computeSlotIndex(metadataUpdate.blockId);
+                    addToConfirmedBlocksBySlot(metadataUpdate.blockId, slot);
+                }
+            }
+        }
+    }
+
+    function onSlotFinalized(slotFinalized: SlotIndex): void {
+        const slotsBefore = Array.from(confirmedBlocksBySlot.keys());
+
+        const slots = [...slotsBefore, slotFinalized];
+
+        const blocks = [];
+        for (const slot of slots) {
+            const blockIds = confirmedBlocksBySlot.get(slot);
+            if (blockIds) {
+                blocks.push(...blockIds);
+            }
+        }
+
+        if (blocks?.length) {
+            blocks.forEach((blockId) => {
+                const selectedColor = getBlockColorByState(themeMode, "finalized");
+                if (selectedColor) {
+                    updateBlockIdToMetadata(blockId, {
+                        state: "finalized",
+                        color: selectedColor,
+                    });
+                    updateBlockColor(blockId, selectedColor);
+                }
+            });
+        }
+
+        for (const slot of slots) {
+            removeConfirmedBlocksSlot(slot);
```

Comment:
It looks nice, I suggest renaming the variables a bit so it is more easy to understand :) 

```typescript
        const previousFinalizedSlots = Array.from(confirmedBlocksBySlot.keys());
        const allFinalizedSlots = [...previousFinalizedSlots, slotFinalized];
        const confirmedBlocks = [];

        for (const slot of allFinalizedSlots) {
            const confirmedBlockIds = confirmedBlocksBySlot.get(slot);

            if (confirmedBlockIds) {
                confirmedBlocks.push(...confirmedBlockIds);
            }
        }

        if (confirmedBlocks?.length) {
            confirmedBlocks.forEach((blockId) => {
                const finalizedBlockColor = getBlockColorByState(themeMode, "finalized");
                if (finalizedBlockColor) {
                    updateBlockIdToMetadata(blockId, {
                        state: "finalized",
                        color: finalizedBlockColor,
                    });
                    updateBlockColor(blockId, finalizedBlockColor);
                }
            });
        }

        for (const slot of allFinalizedSlots) {
            removeConfirmedBlocksSlot(slot);
```